pina/callbacks/adaptive_refinment_callbacks.py

```diff
@@ -96,10 +96,12 @@ def _r3_routine(self, trainer):
         avg = (tot_loss.mean()).to("cpu")
 
         # points to keep
-        old_pts = {}
-        tot_points = 0
+        old_pts = {} # points to retain
+        left_pts = {} # number of points to be uniformly sampled from the respective location
```

Comment:
why do you still need this? If you have `self._const_pts` and `old_pts` then `numb_pts` for each location is `len(self._const_pts[loc])-old_pts[loc]` .

Reply:
This is how the code may look like without computing `left_pts` on fly (and hence, without the `assertion` check)

`def _r3_routine(self, trainer):
        """
        R3 refinement main routine.

        :param Trainer trainer: PINA Trainer.
        """
        # compute residual (all device possible)
        tot_loss, res_loss = self._compute_residual(trainer)
        tot_loss = tot_loss.as_subclass(torch.Tensor)

        # !!!!!! From now everything is performed on CPU !!!!!!

        # average loss
        avg = (tot_loss.mean()).to("cpu")

        # points to keep
        old_pts = {} # points to retain
        for location in self._sampling_locations:
            pts = trainer._model.problem.input_pts[location]
            labels = pts.labels
            pts = pts.cpu().detach().as_subclass(torch.Tensor)
            residuals = res_loss[location].cpu()
            mask = (residuals > avg).flatten()
            if any(mask):  
                pts = (pts[mask]).as_subclass(LabelTensor)
                pts.labels = labels
                old_pts[location] = pts
                

        # sample new points uniformly in the respective location
        for  loc in self._sampling_locations:
            # check if the key is in the dictionary
            if loc in old_pts:
                numb_pts =  self._const_pts[loc] - len(old_pts[loc]) 
            else:
                numb_pts = self._const_pts[loc]
            trainer._model.problem.discretise_domain(
                numb_pts, "random", locations=[loc]
            )

        # adding previous population points
        trainer._model.problem.add_points(old_pts)

        # update dataloader
        trainer._create_or_update_loader()`